File: backend/address_parser.py
"""Address parsing for international formats (AU, US, UK, SA, and more)"""

# Recognized state codes (AU + US)
STATE_CODES = {
    "AL",
    "AK",
    "AZ",
    "AR",
    "CA",
    "CO",
    "CT",
    "DE",
    "FL",
    "GA",
    "HI",
    "ID",
    "IL",
    "IN",
    "IA",
    "KS",
    "KY",
    "LA",
    "ME",
    "MD",
    "MA",
    "MI",
    "MN",
    "MS",
    "MO",
    "MT",
    "NE",
    "NV",
    "NH",
    "NJ",
    "NM",
    "NY",
    "NC",
    "ND",
    "OH",
    "OK",
    "OR",
    "PA",
    "RI",
    "SC",
    "SD",
    "TN",
    "TX",
    "UT",
    "VT",
    "VA",
    "WA",
    "WV",
    "WI",
    "WY",
    "VIC",
    "NSW",
    "QLD",
    "SA",
    "WA",
    "TAS",
    "NT",
    "ACT",
}

# Countries (last part of address) — used to skip
COUNTRIES = {
    "Australia",
    "USA",
    "United States",
    "UK",
    "United Kingdom",
    "Saudi Arabia",
    "Canada",
    "India",
    "UAE",
    "Germany",
    "France",
}


def parse_address(formatted_address):
    """Parse Google formatted address into (address_line, city, state, postcode)"""
    if not formatted_address:
        return "", "", "", ""

    parts = [p.strip() for p in formatted_address.split(",")]
    state = city = postcode = address_line = ""
    state_part_idx = None

    # Step 1: Try to find a recognized state code
    for i, part in enumerate(parts):
        tokens = part.strip().split()
        for s in STATE_CODES:
            if s in tokens:
                state = s
                state_part_idx = i
                for t in reversed(tokens):
                    if t.isdigit() and len(t) >= 4:
                        postcode = t
                        break
                same_part_city = " ".join(
                    t for t in tokens if t != s and not (t.isdigit() and len(t) >= 4)
                )
                if same_part_city:
                    city = same_part_city
                elif i > 0:
                    city = parts[i - 1]
                break
        if state_part_idx is not None:
            break

    # Step 2: Fallback — no state code found (e.g. Saudi Arabia, UK, etc.)
    if state_part_idx is None and len(parts) >= 2:
        # Last part is usually the country — skip it
        last_idx = len(parts) - 1
        if parts[last_idx].strip() in COUNTRIES:
            last_idx -= 1

        # Walk backwards to find the part with a postcode (4+ digits)
        for i in range(last_idx, 0, -1):
            tokens = parts[i].strip().split()
            found_postcode = None
            for t in reversed(tokens):
                if t.isdigit() and len(t) >= 4:
                    found_postcode = t
                    break
            if found_postcode:
                postcode = found_postcode
                city = " ".join(t for t in tokens if t != found_postcode)
                state_part_idx = i
                break

        # If still no postcode, use the second-to-last meaningful part as city
        if state_part_idx is None:
            for i in range(last_idx, 0, -1):
                tokens = parts[i].strip().split()
                all_num = all(t.isdigit() for t in tokens)
                if not all_num and tokens:
                    city = parts[i]
                    state_part_idx = i
                    break

    # Build address_line from parts before city/state
    if state_part_idx is not None and state_part_idx > 0:
        address_line = ", ".join(parts[:state_part_idx])

    return address_line, city, state, postcode

File: lib/places/address-parser.ts
// Address parsing for international formats (AU, US, UK, SA, and more)
// Ported from backend/address_parser.py

const STATE_CODES = new Set([
  "AL", "AK", "AZ", "AR", "CA", "CO", "CT", "DE", "FL", "GA", "HI", "ID",
  "IL", "IN", "IA", "KS", "KY", "LA", "ME", "MD", "MA", "MI", "MN", "MS",
  "MO", "MT", "NE", "NV", "NH", "NJ", "NM", "NY", "NC", "ND", "OH", "OK",
  "OR", "PA", "RI", "SC", "SD", "TN", "TX", "UT", "VT", "VA", "WA", "WV",
  "WI", "WY", "VIC", "NSW", "QLD", "SA", "TAS", "NT", "ACT",
])

const COUNTRIES = new Set([
  "Australia", "USA", "United States", "UK", "United Kingdom",
  "Saudi Arabia", "Canada", "India", "UAE", "Germany", "France",
])

function isPostcode(token: string): boolean {
  return /^\d+$/.test(token) && token.length >= 4
}

export interface ParsedAddress {
  address_line: string
  city: string
  state: string
  postcode: string
}

export function parseAddress(formattedAddress: string): ParsedAddress {
  if (!formattedAddress) {
    return { address_line: "", city: "", state: "", postcode: "" }
  }

  const parts = formattedAddress.split(",").map((p) => p.trim())
  let state = ""
  let city = ""
  let postcode = ""
  let addressLine = ""
  let statePartIdx: number | null = null

  // Step 1: Try to find a recognized state code
  for (let i = 0; i < parts.length; i++) {
    const tokens = parts[i].trim().split(/\s+/).filter(Boolean)
    let matched = false
    for (const token of tokens) {
      if (STATE_CODES.has(token)) {
        state = token
        statePartIdx = i
        for (let t = tokens.length - 1; t >= 0; t--) {
          if (isPostcode(tokens[t])) {
            postcode = tokens[t]
            break
          }
        }
        const samePartCity = tokens
          .filter((t) => t !== token && !isPostcode(t))
          .join(" ")
        if (samePartCity) {
          city = samePartCity
        } else if (i > 0) {
          city = parts[i - 1]
        }
        matched = true
        break
      }
    }
    if (matched) break
  }

  // Step 2: Fallback — no state code found (e.g. Saudi Arabia, UK, etc.)
  if (statePartIdx === null && parts.length >= 2) {
    let lastIdx = parts.length - 1
    if (COUNTRIES.has(parts[lastIdx].trim())) {
      lastIdx -= 1
    }

    for (let i = lastIdx; i > 0; i--) {
      const tokens = parts[i].trim().split(/\s+/).filter(Boolean)
      let foundPostcode: string | null = null
      for (let t = tokens.length - 1; t >= 0; t--) {
        if (isPostcode(tokens[t])) {
          foundPostcode = tokens[t]
          break
        }
      }
      if (foundPostcode) {
        postcode = foundPostcode
        city = tokens.filter((t) => t !== foundPostcode).join(" ")
        statePartIdx = i
        break
      }
    }

    if (statePartIdx === null) {
      for (let i = lastIdx; i > 0; i--) {
        const tokens = parts[i].trim().split(/\s+/).filter(Boolean)
        const allNum = tokens.length > 0 && tokens.every((t) => /^\d+$/.test(t))
        if (!allNum && tokens.length > 0) {
          city = parts[i]
          statePartIdx = i
          break
        }
      }
    }
  }

  if (statePartIdx !== null && statePartIdx > 0) {
    addressLine = parts.slice(0, statePartIdx).join(", ")
  }

  return { address_line: addressLine, city, state, postcode }
}
